### PR TITLE
Create the DirCrawlModelDAO

### PR DESCRIPTION
**Pull request by mcarcaso**
This DAO will crawl the given directory and fill itself with any models, libs, relationships, interfaces, etc. that it finds. This will be an integral part to the build tool.

### CODE REVIEW

File: src/foam/build/DirCrawlModelDAO.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        promises.push(Promise.all(foam.json.references(self.__context__, m)).then(function(p) {
+          var cls = self.lookup(m.class || 'foam.core.Model')
+          return dao.put(cls.create(m));
+        }))
+      };
+
+      context.foam.RELATIONSHIP = function(m) {
+        var r = self.Relationship.create(m);
+        promises.push(Promise.all([
+          self.classloader.load(r.sourceModel),
```
Comment (adamvy) [inline src/foam/build/DirCrawlModelDAO.js]: Why is this necessary?  I don't think the Relationship model will try to resolve source/targetModel anymore until initRelationship() is called.

File: src/foam/build/DirCrawlModelDAO.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        ]).then(function() {
+          return dao.put(r);
+        }))
+      };
+
+      var self = this;
+      var scriptFilesExp = new RegExp(self.unwrappedScripts.join('|'));
+      var blacklistExp = new RegExp(self.blacklist.join('|'));
+      var files = require('child_process')
+        .execSync(`find ${self.srcDir}`)
```
Comment (adamvy) [inline src/foam/build/DirCrawlModelDAO.js]: Oh god.  Let's implement a file tree walk in JS instead.  Something like the posix ftw() function https://linux.die.net/man/3/ftw

File: src/foam/build/DirCrawlModelDAO.js (hunk 1 above)
Comment (mcarcaso) [reply src/foam/build/DirCrawlModelDAO.js]: It's used when figuring out the package which is needed for the ID:
https://github.com/foam-framework/foam2/blob/c7ddf2d90bce6d81247c229f493b4349b3988241/src/foam/dao/Relationship.js#L51

File: src/foam/build/DirCrawlModelDAO.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        .filter(function(o) {
+          return !scriptFilesExp.exec(o) && !blacklistExp.exec(o)
+        })
+
+      var scriptFiles = jsFiles
+        .filter(function(o) {
+          return scriptFilesExp.exec(o) && !blacklistExp.exec(o)
+        })
+
+      // TODO: Do something with java files?
```
Comment (adamvy) [inline src/foam/build/DirCrawlModelDAO.js]: TODO: Add a JavaClass model for manually written java classes.  Parse the package/name from the .java and then store them in the DAO as well.

Also TODO: Do the same for swift.

Comment (mcarcaso) [reply src/foam/build/DirCrawlModelDAO.js]: Done

File: src/foam/build/DirCrawlModelDAO.js (hunk 2 above)
Comment (mcarcaso) [reply src/foam/build/DirCrawlModelDAO.js]: Done

File: src/foam/build/DirCrawlModelDAO.js (hunk 1 above)
Comment (mcarcaso) [reply src/foam/build/DirCrawlModelDAO.js]: Updated the foam.dao.Relationship code to figure out the package/name without having to do a lookup so I was able to remove this.
